File: sdks/jvm/src/main/java/rs/ctx/memory/MemoryClient.java
package rs.ctx.memory;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public class MemoryClient {
    private final MemoryTransport transport;

    protected MemoryClient(MemoryTransport transport) {
        this.transport = transport;
    }

    public static MemoryClient local() {
        return local(LocalCliConfig.builder().build());
    }

    public static MemoryClient local(LocalCliConfig config) {
        return new MemoryClient(new LocalCliAdapter(config));
    }

    public static HostedMemoryClient hosted(HostedConfig config) {
        return new HostedMemoryClient(config);
    }

    public static MemoryClient withTransport(MemoryTransport transport) {
        return new MemoryClient(Objects.requireNonNull(transport, "transport"));
    }

    public StatusResponse status() {
        return new StatusResponse(executeEnvelope("status", list("status", "--json")));
    }

    public InitResponse init() {
        return init(MemoryOptions.init());
    }

    public InitResponse init(MemoryOptions.Init options) {
        MemoryOptions.Init safe = options == null ? MemoryOptions.init() : options;
        List<String> args = mutable("setup", "--json");
        if (safe.progress() != null) {
            args.add("--progress");
            args.add(safe.progress());
        } else {
            args.add("--progress");
            args.add("none");
        }
        if (safe.catalogOnly()) {
            args.add("--catalog-only");
        }
        return new InitResponse(executeEnvelope("init", args));
    }

    public SourcesResponse sources() {
        return new SourcesResponse(executeEnvelope("sources", list("sources", "--json")));
    }

    public ImportResponse importMemory() {
        return importMemory(MemoryOptions.importMemory());
    }

    public ImportResponse importMemory(MemoryOptions.ImportMemory options) {
        return new ImportResponse(executeEnvelope("import", importArgs(options)));
    }

    public ImportResponse sync() {
        return sync(MemoryOptions.importMemory());
    }

    public ImportResponse sync(MemoryOptions.ImportMemory options) {
        return new ImportResponse(executeEnvelope("sync", importArgs(options)));
    }

    public SearchResponse search() {
        return search(MemoryOptions.search());
    }

    public SearchResponse search(String query) {
        return search(MemoryOptions.search().query(query));
    }

    public SearchResponse search(MemoryOptions.Search options) {
        MemoryOptions.Search safe = options == null ? MemoryOptions.search() : options;
        List<String> args = new ArrayList<>();
        args.add("search");
        if (safe.query() != null && !safe.query().isEmpty()) {
            args.add(safe.query());
        }
        args.add("--json");
        if (safe.limit() != null) {
            args.add("--limit");
            args.add(String.valueOf(safe.limit()));
        }
        for (String term : safe.terms()) {
            args.add("--term");
            args.add(term);
        }
        add(args, "--provider", safe.provider());
        add(args, "--workspace", safe.workspace());
        add(args, "--since", safe.since());
        add(args, "--event-type", safe.eventType());
        add(args, "--file", safe.file());
        add(args, "--session", safe.session());
        add(args, "--refresh", safe.refresh());
        if (safe.primaryOnly()) args.add("--primary-only");
        if (safe.includeSubagents()) args.add("--include-subagents");
        if (safe.events()) args.add("--events");
        if (safe.includeCurrentSession()) args.add("--include-current-session");
        return new SearchResponse(executeEnvelope("search", args));
    }

    public ShowEventResponse showEvent(String id, MemoryOptions.ShowEvent options) {
        if (id == null || id.isEmpty()) {
            throw new CtxMemoryException.Validation("event id is required");
        }
        MemoryOptions.ShowEvent safe = options == null ? MemoryOptions.showEvent() : options;
        List<String> args = mutable("show", "event", id, "--format", "json");
        addInt(args, "--before", safe.before());
        addInt(args, "--after", safe.after());
        addInt(args, "--window", safe.window());
        return new ShowEventResponse(executeEnvelope("showEvent", args));
    }

    public ShowEventResponse showEvent(String id) {
        return showEvent(id, MemoryOptions.showEvent());
    }

    public ShowSessionResponse showSession(String id, MemoryOptions.ShowSession options) {
        MemoryOptions.ShowSession safe = options == null ? MemoryOptions.showSession() : options;
        String sessionId = id == null ? safe.id() : id;
        List<String> args = mutable("show", "session");
        if (sessionId != null && !sessionId.isEmpty()) {
            args.add(sessionId);
        } else {
            add(args, "--provider", safe.provider());
            add(args, "--provider-session", safe.providerSessionId());
        }
        add(args, "--mode", safe.mode());
        args.add("--format");
        args.add("json");
        return new ShowSessionResponse(executeEnvelope("showSession", args));
    }

    public ShowSessionResponse showSession(String id) {
        return showSession(id, MemoryOptions.showSession());
    }

    public LocateEventResponse locateEvent(String id) {
        if (id == null || id.isEmpty()) {
            throw new CtxMemoryException.Validation("event id is required");
        }
        return new LocateEventResponse(executeEnvelope("locateEvent", list("locate", "event", id, "--format", "json")));
    }

    public LocateSessionResponse locateSession(String id) {
        return locateSession(MemoryOptions.locateSession().id(id));
    }

    public LocateSessionResponse locateSession(MemoryOptions.LocateSession options) {
        MemoryOptions.LocateSession safe = options == null ? MemoryOptions.locateSession() : options;
        List<String> args = mutable("locate", "session");
        if (safe.id() != null && !safe.id().isEmpty()) {
            args.add(safe.id());
        } else {
            add(args, "--provider", safe.provider());
            add(args, "--provider-session", safe.providerSessionId());
        }
        args.add("--format");
        args.add("json");
        return new LocateSessionResponse(executeEnvelope("locateSession", args));
    }

    public VersionInfo version() {
        return new VersionInfo(transport.name(), transport.ctxVersion());
    }

    public Map<String, Object> versioning() {
        return version().asMap();
    }

    protected MemoryEnvelope execute(String operation, List<String> args) {
        return new MemoryEnvelope(executeEnvelope(operation, args));
    }

    protected Map<String, Object> executeEnvelope(String operation, List<String> args) {
        String stdout = transport.execute(new MemoryOperation(operation, args));
        Map<String, Object> raw;
        try {
            raw = Json.parseObject(stdout);
        } catch (RuntimeException error) {
            Map<String, Object> details = new LinkedHashMap<>();
            details.put("operation", operation);
            details.put("stdout", stdout);
            throw new CtxMemoryException.Protocol("ctx command returned invalid JSON", details, error);
        }
        return MemoryEnvelope.normalize(operation, backend(), raw);
    }

    protected Backend backend() {
        if (transport.name().startsWith("hosted")) {
            return new Backend("hosted", null, null);
        }
        String dataRoot = null;
        if (transport instanceof LocalCliAdapter) {
            dataRoot = ((LocalCliAdapter) transport).config().dataRoot();
        }
        return new Backend("local", dataRoot, null);
    }

    private static List<String> importArgs(MemoryOptions.ImportMemory options) {
        MemoryOptions.ImportMemory safe = options == null ? MemoryOptions.importMemory() : options;
        List<String> args = mutable("import", "--json");
        if (safe.progress() != null) {
            args.add("--progress");
            args.add(safe.progress());
        } else {
            args.add("--progress");
            args.add("none");
        }
        if (safe.all()) args.add("--all");
        add(args, "--provider", safe.provider());
        add(args, "--path", safe.path());
        if (safe.resume()) args.add("--resume");
        return args;
    }

    private static void add(List<String> args, String flag, String value) {
        if (value != null && !value.isEmpty()) {
            args.add(flag);
            args.add(value);
        }
    }

    private static void addInt(List<String> args, String flag, Integer value) {
        if (value != null) {
            args.add(flag);
            args.add(String.valueOf(value));
        }
    }

    private static List<String> list(String... values) {
        List<String> out = mutable(values);
        return Collections.unmodifiableList(out);
    }

    private static List<String> mutable(String... values) {
        List<String> out = new ArrayList<>();
        Collections.addAll(out, values);
        return out;
    }
}

File: sdks/dotnet/src/Ctx.Memory/MemoryClient.cs
using System.Text.Json.Nodes;

namespace Ctx.Memory;

/// <summary>Client for the experimental ctx memory-v1 API.</summary>
public sealed class MemoryClient
{
    private readonly IMemoryTransport _transport;

    public MemoryClient(IMemoryTransport transport)
    {
        _transport = transport;
    }

    public IMemoryTransport Transport => _transport;

    public static MemoryClient Local(LocalMemoryConfig? config = null)
    {
        return new MemoryClient(new LocalCliAdapter(config));
    }

    public static MemoryClient Hosted(HostedMemoryConfig config)
    {
        return new MemoryClient(new HostedAdapter(config));
    }

    public async Task<StatusResponse> StatusAsync(CancellationToken cancellationToken = default)
    {
        var raw = await InvokeAsync("status", ["status", "--json"], cancellationToken).ConfigureAwait(false);
        var envelope = MemoryContract.Envelope("status", _transport.Backend(raw), "status", MemoryContract.NormalizeStatus(raw));
        return new StatusResponse(envelope);
    }

    public async Task<InitResponse> InitAsync(InitOptions? options = null, CancellationToken cancellationToken = default)
    {
        options ??= new InitOptions();
        var args = new List<string> { "setup", "--json" };
        AddOption(args, "--progress", options.Progress);
        if (options.CatalogOnly)
        {
            args.Add("--catalog-only");
        }

        var raw = await InvokeAsync("init", args, cancellationToken).ConfigureAwait(false);
        var envelope = MemoryContract.Envelope("init", _transport.Backend(raw), "status", MemoryContract.NormalizeStatus(raw));
        return new InitResponse(envelope);
    }

    public async Task<SourcesResponse> SourcesAsync(CancellationToken cancellationToken = default)
    {
        var raw = await InvokeAsync("sources", ["sources", "--json"], cancellationToken).ConfigureAwait(false);
        var envelope = MemoryContract.Envelope("sources", _transport.Backend(raw), "sources", MemoryContract.NormalizeSources(raw));
        return new SourcesResponse(envelope);
    }

    public async Task<ImportResponse> ImportMemoryAsync(ImportOptions? options = null, CancellationToken cancellationToken = default)
    {
        return await ImportLikeAsync("import", options, cancellationToken).ConfigureAwait(false);
    }

    public async Task<ImportResponse> ImportAsync(ImportOptions? options = null, CancellationToken cancellationToken = default)
    {
        return await ImportMemoryAsync(options, cancellationToken).ConfigureAwait(false);
    }

    public async Task<ImportResponse> SyncAsync(ImportOptions? options = null, CancellationToken cancellationToken = default)
    {
        return await ImportLikeAsync("sync", options, cancellationToken).ConfigureAwait(false);
    }

    public async Task<SearchResponse> SearchAsync(SearchOptions? options = null, CancellationToken cancellationToken = default)
    {
        options ??= new SearchOptions();
        var args = new List<string> { "search" };
        if (!string.IsNullOrWhiteSpace(options.Query))
        {
            args.Add(options.Query);
        }
        foreach (var term in options.Terms ?? [])
        {
            AddOption(args, "--term", term);
        }
        if (options.Limit is > 0)
        {
            AddOption(args, "--limit", options.Limit.Value.ToString(System.Globalization.CultureInfo.InvariantCulture));
        }
        AddOption(args, "--provider", options.Provider);
        AddOption(args, "--workspace", options.Workspace);
        AddOption(args, "--since", options.Since);
        if (options.PrimaryOnly)
        {
            args.Add("--primary-only");
        }
        if (options.IncludeSubagents)
        {
            args.Add("--include-subagents");
        }
        AddOption(args, "--event-type", options.EventType);
        AddOption(args, "--file", options.File);
        AddOption(args, "--session", options.Session);
        if (options.Events)
        {
            args.Add("--events");
        }
        AddOption(args, "--refresh", options.Refresh);
        if (options.IncludeCurrentSession)
        {
            args.Add("--include-current-session");
        }
        args.Add("--json");

        var raw = await InvokeAsync("search", args, cancellationToken).ConfigureAwait(false);
        var envelope = MemoryContract.Envelope("search", _transport.Backend(raw), "search", MemoryContract.NormalizeSearch(raw));
        return new SearchResponse(envelope);
    }

    public async Task<ShowEventResponse> ShowEventAsync(
        string eventId,
        ShowEventOptions? options = null,
        CancellationToken cancellationToken = default)
    {
        RequireValue(eventId, "event id");
        options ??= new ShowEventOptions();

        var args = new List<string> { "show", "event", eventId, "--format", "json" };
        AddNumberOption(args, "--before", options.Before);
        AddNumberOption(args, "--after", options.After);
        AddNumberOption(args, "--window", options.Window);

        var raw = await InvokeAsync("showEvent", args, cancellationToken).ConfigureAwait(false);
        var envelope = MemoryContract.Envelope("showEvent", _transport.Backend(raw), "event", MemoryContract.NormalizeEvent(raw));
        return new ShowEventResponse(envelope);
    }

    public async Task<ShowSessionResponse> ShowSessionAsync(
        string sessionId,
        ShowSessionOptions? options = null,
        CancellationToken cancellationToken = default)
    {
        options = (options ?? new ShowSessionOptions()) with { Id = sessionId };
        return await ShowSessionAsync(options, cancellationToken).ConfigureAwait(false);
    }

    public async Task<ShowSessionResponse> ShowSessionAsync(
        ShowSessionOptions options,
        CancellationToken cancellationToken = default)
    {
        var args = BuildSessionLookupArgs("show", "session", options);
        AddOption(args, "--mode", options.Mode);
        args.Add("--format");
        args.Add("json");

        var raw = await InvokeAsync("showSession", args, cancellationToken).ConfigureAwait(false);
        var envelope = MemoryContract.Envelope("showSession", _transport.Backend(raw), "session", MemoryContract.NormalizeSession(raw));
        return new ShowSessionResponse(envelope);
    }

    public async Task<LocateEventResponse> LocateEventAsync(string eventId, CancellationToken cancellationToken = default)
    {
        RequireValue(eventId, "event id");
        var raw = await InvokeAsync(
            "locateEvent",
            ["locate", "event", eventId, "--format", "json"],
            cancellationToken).ConfigureAwait(false);
        var envelope = MemoryContract.Envelope("locateEvent", _transport.Backend(raw), "location", MemoryContract.NormalizeLocation(raw));
        return new LocateEventResponse(envelope);
    }

    public async Task<LocateSessionResponse> LocateSessionAsync(string sessionId, CancellationToken cancellationToken = default)
    {
        return await LocateSessionAsync(new SessionLookupOptions { Id = sessionId }, cancellationToken).ConfigureAwait(false);
    }

    public async Task<LocateSessionResponse> LocateSessionAsync(
        SessionLookupOptions options,
        CancellationToken cancellationToken = default)
    {
        var args = BuildSessionLookupArgs("locate", "session", options);
        args.Add("--format");
        args.Add("json");

        var raw = await InvokeAsync("locateSession", args, cancellationToken).ConfigureAwait(false);
        var envelope = MemoryContract.Envelope("locateSession", _transport.Backend(raw), "location", MemoryContract.NormalizeLocation(raw));
        return new LocateSessionResponse(envelope);
    }

    public async Task<VersionInfo> VersionAsync(CancellationToken cancellationToken = default)
    {
        var ctxVersion = await _transport.GetCtxVersionAsync(cancellationToken).ConfigureAwait(false);
        return new VersionInfo(
            CtxMemoryVersions.SdkVersion,
            CtxMemoryVersions.ContractVersion,
            _transport.Name,
            ctxVersion);
    }

    public async Task<VersionInfo> VersioningAsync(CancellationToken cancellationToken = default)
    {
        return await VersionAsync(cancellationToken).ConfigureAwait(false);
    }

    private async Task<ImportResponse> ImportLikeAsync(
        string operation,
        ImportOptions? options,
        CancellationToken cancellationToken)
    {
        options ??= new ImportOptions();
        var args = new List<string> { "import", "--json" };
        AddOption(args, "--progress", options.Progress);
        if (options.All)
        {
            args.Add("--all");
        }
        AddOption(args, "--provider", options.Provider);
        AddOption(args, "--path", options.Path);
        if (options.Resume)
        {
            args.Add("--resume");
        }

        var raw = await InvokeAsync(operation, args, cancellationToken).ConfigureAwait(false);
        var envelope = MemoryContract.Envelope(operation, _transport.Backend(raw), "import", MemoryContract.NormalizeImport(raw));
        return new ImportResponse(envelope);
    }

    private async Task<JsonObject> InvokeAsync(
        string operation,
        IReadOnlyList<string> args,
        CancellationToken cancellationToken)
    {
        var raw = await _transport.ExecuteJsonAsync(operation, args, cancellationToken).ConfigureAwait(false);
        MemoryContract.EnsureSupportedSchema(raw, operation);
        return raw;
    }

    private static List<string> BuildSessionLookupArgs(string command, string kind, SessionLookupOptions options)
    {
        var args = new List<string> { command, kind };
        if (!string.IsNullOrWhiteSpace(options.Id))
        {
            args.Add(options.Id);
        }
        else if (!string.IsNullOrWhiteSpace(options.ProviderSessionId))
        {
            AddOption(args, "--provider", options.Provider);
            AddOption(args, "--provider-session", options.ProviderSessionId);
        }
        else
        {
            throw new CtxMemoryValidationException($"{kind} lookup requires an id or provider session id");
        }
        return args;
    }

    private static void RequireValue(string value, string name)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            throw new CtxMemoryValidationException($"{name} is required");
        }
    }

    private static void AddOption(List<string> args, string flag, string? value)
    {
        if (!string.IsNullOrWhiteSpace(value))
        {
            args.Add(flag);
            args.Add(value);
        }
    }

    private static void AddNumberOption(List<string> args, string flag, int? value)
    {
        if (value is > 0)
        {
            args.Add(flag);
            args.Add(value.Value.ToString(System.Globalization.CultureInfo.InvariantCulture));
        }
    }
}
